src/main/java/at/hannibal2/skyhanni/features/garden/pests/PestSpawnTimer.kt

```diff
@@ -1,39 +1,240 @@
+
     var lastSpawnTime = SimpleTimeMark.farPast()
 
+    private var longestCropBrokenTime: Duration = 0.seconds
+
+    private var pestSpawned = false
+
+    private var hasWarned = false
+
+    private var maxPests = false
+
+    private var ready = false
+
+    @HandleEvent(onlyOnIsland = IslandType.GARDEN)
+    fun onWidgetUpdate(event: WidgetUpdateEvent) {
+        if (!event.isWidget(TabWidget.PESTS)) return
+
+        pestCooldownPattern.firstMatcher(event.widget.lines) {
+            val minutes = groupOrNull("minutes")?.formatInt()
+            val seconds = groupOrNull("seconds")?.formatInt()
+            ready = hasGroup("ready")
+            maxPests = hasGroup("maxPests")
+
+            if (ready || maxPests) return
+
+            if (minutes == null && seconds == null) return
+
+            val tablistCooldownEnd = SimpleTimeMark.now() + (minutes?.minutes ?: 0.seconds) + (seconds?.seconds ?: 0.seconds)
+
+            if (shouldSetCooldown(tablistCooldownEnd, minutes, seconds)) {
+
+                // hypixel sometimes rounds down times, we'll assume times are rounded down if seconds are null and add a minute
+
+                pestCooldownEndTime = if (seconds == null) {
+                    tablistCooldownEnd + 1.minutes
+                } else {
+                    tablistCooldownEnd
+                }
+
+                if (pestSpawned) {
+                    hasWarned = false
+                    pestSpawned = false
+                }
+            }
+        }
+    }
+
     @HandleEvent
     fun onPestSpawn(event: PestSpawnEvent) {
+        val spawnTime = lastSpawnTime.passedSince()
+
+        if (!lastSpawnTime.isFarPast()) {
+            if (longestCropBrokenTime.inWholeSeconds.toInt() <= config.averagePestSpawnTimeout) {
+                pestSpawnTimes.add(spawnTime.inWholeSeconds.toInt())
+                ChatUtils.debug("Added pest spawn time ${spawnTime.format()}")
+            }
+
+            if (config.pestSpawnChatMessage) {
+                ChatUtils.chat("Pests spawned in §b${spawnTime.format()}")
+            }
+        }
+
+        pestSpawned = true
+
+        longestCropBrokenTime = 0.seconds
+
         lastSpawnTime = SimpleTimeMark.now()
+
     }
 
     @HandleEvent
     fun onRenderOverlay(event: GuiRenderEvent.GuiOverlayRenderEvent) {
         if (!isEnabled()) return
-        if (config.onlyWithVacuum && !PestApi.hasVacuumInHand()) return
 
-        val display = if (lastSpawnTime.isFarPast()) {
+        if (config.onlyWithVacuum xor config.onlyWithFarmingTool) {
+            if (config.onlyWithFarmingTool && !GardenApi.hasFarmingToolInHand()) return
+            if (config.onlyWithVacuum && !PestApi.hasVacuumInHand()) return
+        } else if (config.onlyWithFarmingTool && config.onlyWithVacuum) {
+            if (!GardenApi.hasFarmingToolInHand() && !PestApi.hasVacuumInHand()) return
+        }
+
+        config.position.renderRenderables(drawDisplay(), posLabel = "Pest Spawn Timer")
```

Comment:
please cache the display. this redraws it every frame. update the diplay info only once per second (idk why more often would be helpful here)